### PR TITLE
Critical Generic comparison bug found

### PR DESCRIPTION
**Pull request by korken89**
I found a `Generic` comparison bug that is quite critical.

### CODE REVIEW

**Comment by korken89** [comment]
The old checking code truncated the checked values causing many comparisons to become `0.cmp(&0)`, this is now fixed.

**Comment by andresv** [comment]
I can confirm that this fixes panic in such rtic idle loop:

```rust
let mut t = monotonics::now();
loop {
    let now = monotonics::now();
    if now - t >= Milliseconds(500).into() {
        t = monotonics::now();
        let _ = ctx.local.led.toggle();
    }
}
```

Without this it panics in `now - t`:
```
7764 ERROR panicked at 'Sub failed', /Users/andres/.cargo/registry/src/github.com-1ecc6299db9ec823/embedded-time-0.12.1/./src/instant.rs:454:13
```

It is critical to get it merged.

**Comment by korken89** [comment]
Hmm, while this works the 3 multiplications can easily overflow.
One should probably run some form of GCD between terms before performing multiplications, or some other smart way of performing the check.

**Comment by PTaylor-us** [comment]
Wow, yeah. This was poorly implemented on my part. 

Unless I'm mistaken, you're simply finding and comparing the numerators that would result from a cross-multiplication. I agree with you about the easy overflow on the multiplication. I am going to add the GCD reductions using the existing [Stein's Algorithm](https://www.geeksforgeeks.org/steins-algorithm-for-finding-gcd/) GCD implementation in the `num-integer` crate.

**Comment by PTaylor-us** [comment]
I'm still poking around a bit. I think an initial implementation should follow the patterns present in the _named_ Duration/Rate implementations. They convert the value with the larger scaling factor into one with the smaller scaling factory (e.g. Seconds -> Milliseconds). If the conversion fails, than that shows us it's greater. The step would be implementing a `Generic<T>::try_into_scaling_factor()` that will attempt to make the conversion in the same way as the `TryInto<Milliseconds<T>> for Seconds<T>` (implemented as `TryFrom`s). There are certainly more efficient ways of doing it, but it can be optimized as needed after we have the desired functionality.

**Comment by PTaylor-us** [comment]
@korken89 Please take a look at my changes. It follows virtual the same pattern as the _named_ duration `PartialOrd`. In addition to the normal tests, I tested it with a stress test, looping with random numbers.

File: tests/duration.rs [hunk 1]
```diff
@@ -56,6 +56,53 @@ fn comparisons() {
...
+    );
+    assert!(
+        Generic::new(100u32, Fraction::new(1, 1000)) <= Generic::new(11u32, Fraction::new(1, 100))
+    );
+    assert!(
+        Generic::new(200u32, Fraction::new(1, 1000)) >= Generic::new(10u32, Fraction::new(1, 100))
+    );
+    assert!(
+        Generic::new(200u32, Fraction::new(1, 1000)) >= Generic::new(10u32, Fraction::new(1, 100))
+    );
```
Comment (str4d) [inline tests/duration.rs:75]: These two test cases are identical. These should either be de-duplicated, one of these changed to cover a different part of the comparison space, or a comment added as to why the duplicate is present.

File: src/time_int.rs [hunk 2]
```diff
@@ -32,6 +32,14 @@ pub trait TimeInt:
...
         self.checked_mul_fraction(&fraction.recip())
     }
+
+    /// Moves an integer into a comparable base for checking
+    fn checked_same_base(&self, fraction: &Fraction, rhs_fraction: &Fraction) -> Option<Self> {
+       let a_n = *fraction.numerator();
+       let b_d = *rhs_fraction.denominator();
+
+       self.checked_mul(&(b_d.into()))?.checked_mul(&(a_n.into()))
+    }
```
Comment (str4d) [inline src/time_int.rs:42]: This is now unused in the PR after the refactor:
```suggestion
```

File: tests/duration.rs [hunk 3]
```diff
@@ -56,6 +56,53 @@ fn comparisons() {
...
+                    }
+                }
+            }
+        }
+        println!("rands: {:#?}", rands);
+        assert!(
+            Generic::new(rands[0], Fraction::new(rands[1], rands[2]))
+                != Generic::new(rands[3], Fraction::new(rands[4], rands[5]))
+        );
+    } */
```
Comment (str4d) [inline tests/duration.rs:105]: It might be useful to rewrite this using the [`proptest` crate](https://crates.io/crates/proptest), so you retain the "random checks" testing, while also having bounded runtimes for tests, and reductions of failures to simplified cases.

File: src/duration.rs [hunk 4]
```diff
@@ -443,14 +443,42 @@ pub struct Generic<T> {
     scaling_factor: Fraction,
 }
 
+impl<T: TimeInt> Generic<T> {
+    /// Try to create a new, equivalent `Generic` with the given _scaling factor_
```
Comment (str4d) [inline src/duration.rs:447]: Just to help make what is going on here a little clearer:
```suggestion
    /// Try to create a new, equivalent `Generic` with the given _scaling factor_.
    ///
    /// To minimise errors due to truncation rounding, the new integer component is
    /// calculated as `self.integer * (self.scaling_factor / scaling_factor)`.
```

File: src/duration.rs [hunk 5]
```diff
@@ -443,14 +443,42 @@ pub struct Generic<T> {
...
-        Some(
-            self.integer
-                .checked_mul_fraction(&self.scaling_factor)?
-                .cmp(&rhs.integer.checked_mul_fraction(&rhs.scaling_factor)?),
-        )
+        if self.scaling_factor == rhs.scaling_factor {
+            Some(self.integer.cmp(&rhs.integer))
+        } else if self.scaling_factor < rhs.scaling_factor {
+            // convert to the smaller scaling factor (rhs -> self)
+            // if conversion fails, we know self is less than rhs
```
Comment (str4d) [inline src/duration.rs:469]: It feels like this will be true most of the time, but I'm not entirely convinced that there aren't edge cases where `rhs.scaling_factor / self.scaling_factor < self.integer / rhs.integer` (so `self > rhs`), and `rhs` is close enough to the maximum bound on `T` that it will overflow. In particular, `TimeInt::checked_mul_fraction` is being called with `rhs.integer` and a fraction that is greater than 1, and is implemented by first multiplying by the numerator (which could overflow) and then dividing by the denominator (which if not for the overflow could bring the result back in range and below `self.integer`). It would be great to have tests that exercise these near-bounds edge cases.

Comment (korken89) [reply src/duration.rs:469]: I have stopped contributing here and created [`fugit`](https://crates.io/crates/fugit) instead.

File: src/duration.rs (hunk 4 above)
Comment (korken89) [reply src/duration.rs:447]: I have stopped contributing here and created [`fugit`](https://crates.io/crates/fugit) instead.

File: tests/duration.rs (hunk 3 above)
Comment (korken89) [reply tests/duration.rs:105]: I have stopped contributing here and created [`fugit`](https://crates.io/crates/fugit) instead.

File: src/time_int.rs (hunk 2 above)
Comment (korken89) [reply src/time_int.rs:42]: I have stopped contributing here and created [`fugit`](https://crates.io/crates/fugit) instead.

File: tests/duration.rs (hunk 1 above)
Comment (korken89) [reply tests/duration.rs:75]: I have stopped contributing here and created [`fugit`](https://crates.io/crates/fugit) instead.
